src/libfsm/print/ir.c

```diff
@@ -106,13 +104,29 @@ make_groups(const struct fsm *fsm, fsm_state_t state,
 
 	assert(n > 0);
 
+	/* Sort and combine consecutive states. */
+	qsort(ranges, n, sizeof *ranges, range_cmp);
+	{
+		size_t src, dst = 0;
+		for (src = 1; src < n; src++) {
+			if (ranges[src].to == ranges[dst].to
+			    && ranges[src].start == ranges[dst].end + 1) {
+				ranges[dst].end = ranges[src].end;
+			} else {
+				assert(dst < src);
+				dst++;
+				memcpy(&ranges[dst], &ranges[src],
+				    sizeof(ranges[dst]));
+			}
+		}
+		n = dst + 1;
+	}
+
```

Comment:
Would you want to move this up to an earlier stage, when we construct the IR?

If there are entries we can combine, everything would benefit (at the cost of the sort, of course. but that's fine). And I'd like the IR printing to represent exactly what the structs contain, to make debugging less confusing.

Reply:
If we can do so unambiguously, earlier would be better. Should it be part of this PR or separate?

Reply:
I just switched this to ordered iteration. I decided to leave the range consolidation in; it may not longer be necessary, but I'm not sure.